x-pack/metricbeat/module/gcp/metrics/dataproc/metadata.go

```diff
@@ -0,0 +1,208 @@
+	organizationID    string
+	organizationName  string
+	regions           []string
+	collectUserLabels bool
+	opt               []option.ClientOption
+	// NOTE: clusters holds data used for all metrics collected in a given period
+	// this avoids calling the remote endpoint for each metric, which would take a long time overall
+	clusters map[string]*dataproc.Cluster
+	logger   *logp.Logger
+}
+
+// Metadata implements googlecloud.MetadataCollector to the known set of labels from a Dataproc TimeSeries single point of data.
+func (s *metadataCollector) Metadata(ctx context.Context, resp *monitoringpb.TimeSeries) (gcp.MetadataCollectorData, error) {
+	stackdriverLabels := gcp.NewStackdriverMetadataServiceForTimeSeries(resp, s.organizationID, s.organizationName, s.projectName)
+
+	metadataCollectorData, err := stackdriverLabels.Metadata(ctx, resp)
+	if err != nil {
+		return gcp.MetadataCollectorData{}, err
+	}
+
+	if resp.Resource != nil && resp.Resource.Labels != nil {
+		_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceIDKey, resp.Resource.Labels["cluster_uuid"])
+	}
+
+	_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudInstanceNameKey, resp.Resource.Labels["cluster_name"])
+
+	if s.collectUserLabels {
+		metadata, err := s.instanceMetadata(ctx, s.instanceID(resp), s.instanceRegion(resp))
+		if err != nil {
+			return gcp.MetadataCollectorData{}, err
+		}
+
+		if metadata.machineType != "" {
+			lastIndex := strings.LastIndex(metadata.machineType, "/")
+			_, _ = metadataCollectorData.ECS.Put(gcp.ECSCloudMachineTypeKey, metadata.machineType[lastIndex+1:])
+		}
+
+		metadata.Metrics = metadataCollectorData.Labels[gcp.LabelMetrics]
+		metadata.System = metadataCollectorData.Labels[gcp.LabelSystem]
+
+		if metadata.User != nil {
+			metadataCollectorData.Labels[gcp.LabelUser] = metadata.User
+		}
+	}
+
+	return metadataCollectorData, nil
+}
+
+// instanceMetadata returns the labels of an instance
+func (s *metadataCollector) instanceMetadata(ctx context.Context, instanceID, region string) (*dataprocMetadata, error) {
+	cluster, err := s.instance(ctx, instanceID)
+	if err != nil {
+		return nil, fmt.Errorf("error trying to get data from instance '%s' %w", instanceID, err)
+	}
+
+	metadata := &dataprocMetadata{
+		clusterID: instanceID,
+		region:    region,
+	}
+
+	if cluster == nil {
+		s.logger.Debugf("couldn't get instance '%s' call ListInstances API", instanceID)
+		return metadata, nil
+	}
+
+	if cluster.ClusterName != "" {
+		parts := strings.Split(cluster.ClusterName, "/")
+		metadata.clusterName = parts[len(parts)-1]
+	}
+
+	if cluster.Labels != nil {
+		metadata.User = cluster.Labels
+	}
+
+	return metadata, nil
+}
+
+// instance returns data from an instance ID from the cache.
+func (s *metadataCollector) instance(ctx context.Context, instanceID string) (*dataproc.Cluster, error) {
+	s.getInstances(ctx)
+
+	instance, ok := s.clusters[instanceID]
+	if ok {
+		return instance, nil
+	}
+
+	s.clusters = make(map[string]*dataproc.Cluster)
```

Comment:
i don't follow this branch. we populate the cache for clusters for all configured regions, but then if the current instance is missing we invalidate the entire cache and return an almost empty metadata object (line 116)? 

is that the correct behaviour?

Reply:
At line 116 we start populating the metadata with specific cluster details `ClusterName` and `Labels`. If the cluster is not found in the cache we simply return the metadata we currently have which is the cluster id and the region, both are retrieved from the labels.

IIRC the point of this line 140 was to avoid having stale data, in `getInstances` there is a check if clusters are present in the cache, if yes it returns. If no cluster is found in the cache we clean it to start the collection again.

Reply:
got it. but the collection only starts again on the next periodic run? 

under what other circumstances is the cache refreshed? what happens if a user adds a label to an instance? do you have to restart metricbeat for it to pick it up?

Reply:
yes the collection would start at the next period defined in the config e.g. `1m`

that's the only time the cache gets refreshed in this case, when a cluster is not found. 

> if a user adds a label to an instance? do you have to restart metricbeat for it to pick it up?

Yeah. The alternative is to never clean that cache (map), that way we collect metadata at each period interval defined in config and the chances of having stale data will be smaller.